File: src/main/java/org/cimsbioko/search/IndividualCursorDocumentSource.java
package org.cimsbioko.search;

import android.database.Cursor;

import static org.cimsbioko.search.Utils.extractDissimilarNames;
import static org.cimsbioko.search.Utils.extractUniquePhones;
import static org.cimsbioko.search.Utils.join;

class IndividualCursorDocumentSource extends SimpleCursorDocumentSource {

    private String nameColumn, phoneColumn;

    IndividualCursorDocumentSource(Cursor c, String nameCol, String phoneCol) {
        super(c);
        nameColumn = nameCol;
        phoneColumn = phoneCol;
    }

    @Override
    public String getFieldValue(int index) {
        String fieldName = getFieldName(index);
        if (nameColumn.equals(fieldName)) {
            return getNameValue(index);
        } else if (phoneColumn.equals(fieldName)) {
            return getPhoneValue(index);
        } else {
            return super.getFieldValue(index);
        }
    }

    private String getPhoneValue(int index) {
        String rawValue = cursor.getString(index);
        if (rawValue != null) {
            return join(extractUniquePhones(rawValue), " ");
        }
        return null;
    }

    private String getNameValue(int index) {
        String rawValue = cursor.getString(index);
        if (rawValue != null) {
            return join(extractDissimilarNames(rawValue), " ");
        }
        return null;
    }


}

File: src/main/kotlin/org/cimsbioko/search/IndividualCursorDocumentSource.kt
package org.cimsbioko.search

import android.database.Cursor
import org.cimsbioko.search.Utils.extractDissimilarNames
import org.cimsbioko.search.Utils.extractUniquePhones
import org.cimsbioko.search.Utils.join

internal class IndividualCursorDocumentSource(
        c: Cursor,
        private val nameColumn: String,
        private val phoneColumn: String
) : SimpleCursorDocumentSource(c) {
    override fun getFieldValue(index: Int) = when (getFieldName(index)) {
        nameColumn -> cursor.getString(index)?.let { join(extractDissimilarNames(it), " ") }
        phoneColumn -> cursor.getString(index)?.let { join(extractUniquePhones(it), " ") }
        else -> super.getFieldValue(index)
    }
}